python/src/equistore/operations/reduce_over_samples.py

```diff
@@ -0,0 +1,168 @@
+import numpy as np
+
+
+from equistore import Labels, TensorBlock, TensorMap
+
+from . import _dispatch
+from typing import List
+
+
+def _reduce_over_samples_block(
+    block: TensorBlock, sample_names: List[str], reduction: str
+) -> TensorBlock:
+    """Create a new :py:class:`TensorBlock` summing the ``properties`` among
+    the selected ``samples``.
+    The output :py:class:`TensorBlocks` have the same components of the input one.
+    :param block: -> input block
+    :param sample_names: -> names of samples to sum
+    """
+    for sample in sample_names:
+        assert (
+            sample in block.samples.names
+        ), "the values in sample_names should be contained in block.samples.names"
+
+    if reduction not in [
+        "sum",
+        "mean",
+    ]:
+        raise ValueError(
+            '_reduce_over_samples_block supports only reduction="sum" or "mean",'
+            f'"{reduction}" was passed'
+        )
+
+    # get the indices of the selected sample
+    sample_selected = [block.samples.names.index(sample) for sample in sample_names]
+    # reshaping the samples in a 2D array
+    samples = block.samples.view(dtype=np.int32).reshape(block.samples.shape[0], -1)
+    s, index = np.unique(samples[:, sample_selected], return_inverse=True, axis=0)
+    values_result = _dispatch.zeros_like(
+        block.values, shape=(s.shape[0],) + block.values.shape[1:]
+    )
+
+    if reduction == "mean":
+        bincount = _dispatch.bincount(index)
```

Comment:
to take care of the mean in a way that works both with numpy and torch I used this bitcount thing.
I do not know if it is the best but there are no python loop

Reply:
that looks good!